### PR TITLE
Enforce Editor Lock for Write, Append, and Upload Operations in DataNode

### PR DESCRIPTION
**Pull request by PrakharJain1509**
### PR Description

**Title:** Enforce Editor Lock for Write, Append, and Upload Operations in DataNode

#### Summary
This PR addresses the issue #2017 where a `DataNode` could be modified by other editors despite being locked by one. It ensures that once a `DataNode` is locked with an `editor_id`, no other editor can perform **write**, **append**, or **upload** operations until the lock is released.

#### Changes Implemented
- Modified the **`write()`**, **`append()`**, and **`_upload()`** methods:
  - Enforced locking mechanism to prevent unauthorized modifications by other editors.
  - Raised `DataNodeIsBeingEdited` exception when a different editor attempts to modify the locked `DataNode`.
  - Allowed orchestrator writes without `editor_id` even during an active lock.
- **Updated unit tests** in:
  - `test_data_node.py` to validate the new behavior.
- Ensured **90% code coverage** on all modified methods.

#### Acceptance Criteria Met
- Locking mechanism for **write**, **append**, and **upload** is now enforced as expected.
- Orchestrator writes remain allowed without `editor_id`.
- All relevant tests pass with the required code coverage.

---

This PR ensures consistent locking behavior across all relevant operations in line with the documentation and expected usage.

### CODE REVIEW

**Comment by PrakharJain1509** [comment]
Hey, @FlorianJacta and @bobbyshermi 
I have created this PR addressing all the changes asked, Please review it and let me know if any changes are required.

File: taipy/core/data/in_memory.py [hunk 1]
```diff
@@ -98,3 +98,16 @@ def _read(self):
 
     def _write(self, data):
         in_memory_storage[self.id] = data
+
+    def _append(self, data):
+        """Append data to the existing data in the in-memory storage."""
+        if self.id not in in_memory_storage:
+            in_memory_storage[self.id] = []
+
+        in_memory_storage[self.id].append(data)  # Append new data
```
Comment (bobbyshermi) [inline taipy/core/data/in_memory.py:109]: What if the previous data of the data node has no `append` method. 
In other words, what if `in_memory_storage[self.id]` has no `append` method?

**Review by bobbyshermi — CHANGES_REQUESTED** [review]

File: taipy/core/data/in_memory.py (hunk 1 above)
Comment (PrakharJain1509) [reply taipy/core/data/in_memory.py:109]: Ok I added a if instance condition to resolve that. Please review it and let me know if any changes required.

**Comment by PrakharJain1509** [comment]
@bobbyshermi Did you review it sir?

**Comment by PrakharJain1509** [comment]
@bobbyshermi Did you review it sir?

**Comment by PrakharJain1509** [comment]
@bobbyshermi @FlorianJacta @jrobinAV 
Please let me know if any changes are required?

**Comment by PrakharJain1509** [comment]
@bobbyshermi @jrobinAV @FlorianJacta Please inform me if this issue is open or closed?

**Comment by PrakharJain1509** [comment]
> @bobbyshermi @jrobinAV @FlorianJacta Please inform me if this issue is open or closed?
@bobbyshermi  @jrobinAV  @FlorianJacta 
???

**Comment by jrobinAV** [comment]
@PrakharJain1509 The issue is still open, yes, and your PR is in the review process. Thank you for your patience. We will get back to you very soon.

File: taipy/core/data/data_node.py [hunk 2]
```diff
@@ -442,9 +444,16 @@ def write(self, data, job_id: Optional[JobId] = None, **kwargs: Dict[str, Any]):
         """
         from ._data_manager_factory import _DataManagerFactory
 
+        if self.edit_in_progress and editor_id and self.editor_id != editor_id:
```
Comment (jrobinAV) [inline taipy/core/data/data_node.py:454]: If I understand correctly, this covers the case where the data node is already locked by a different editor than the one attempting to write it. We also need to check that the `self.editor_expiration_date` (if it exists) is not past.

**Review by toan-quach — DISMISSED** [review]
Thanks for your contribution! The changes look good to me, but tests are failing. Please help take a look and fix it 😃

File: tests/core/data/test_data_node.py [hunk 3]
```diff
@@ -745,3 +745,33 @@ def test_change_data_node_name(self):
         # This new syntax will be the only one allowed: https://github.com/Avaiga/taipy-core/issues/806
         dn.properties["name"] = "baz"
         assert dn.name == "baz"
+
```
Comment (jrobinAV) [inline tests/core/data/test_data_node.py:754]: I believe we are missing a few test cases here:

- An editor `editor_id_1` locked the data node, but the expiration date passed. Another "manual" editor, `editor_id_2,` should succeed in writing the data node.
- An editor `editor_id_1` locked the data node, but the expiration date passed. Another "manual" editor, `editor_id_2,` should succeed in appending the data node.
- The orchestrator locks the data node. A "manual" editor should fail to write the data node.
- The orchestrator locks the data node. The Orchestrator should succeed in writing the data node.
The orchestrator locks the data node. A "manual" editor should fail to append it.

File: taipy/core/data/data_node.py [hunk 4]
```diff
@@ -442,9 +444,16 @@ def write(self, data, job_id: Optional[JobId] = None, **kwargs: Dict[str, Any]):
...
+            raise DataNodeIsBeingEdited(self.id, self.editor_id)
+
+        if not editor_id and self.edit_in_progress:
+            print("Orchestrator writing without editor_id")
```
Comment (jrobinAV) [inline taipy/core/data/data_node.py:458]: This covers the case where the Orchestrator attempts to write a data node locked by an editor. We can eventually add an info log using the `_logger` attribute, but we don't want any print in the production code. Please remove it.

File: taipy/core/data/data_node.py [hunk 5]
```diff
@@ -442,9 +444,16 @@ def write(self, data, job_id: Optional[JobId] = None, **kwargs: Dict[str, Any]):
...
+
         self._write(data)
-        self.track_edit(job_id=job_id, **kwargs)
-        self.unlock_edit()
+        self.last_edit_date = datetime.now()
```
Comment (jrobinAV) [inline taipy/core/data/data_node.py:461]: The `last_edit_date` setting is already done in the `track_edit` method called right after. Please revert this line.

File: taipy/core/data/data_node.py [hunk 6]
```diff
@@ -471,6 +480,10 @@ def lock_edit(self, editor_id: Optional[str] = None):
         Arguments:
             editor_id (Optional[str]): The editor's identifier.
         """
+
+        if self._editor_expiration_date and datetime.now() > self._editor_expiration_date:
+            self.unlock_edit()
+
```
Comment (jrobinAV) [inline taipy/core/data/data_node.py:508]: I don't see what case we want to cover here. For me, the next condition already covers this case. Can you elaborate? Otherwise, we should revert it.

File: taipy/core/data/data_node.py [hunk 7]
```diff
@@ -442,9 +444,16 @@ def write(self, data, job_id: Optional[JobId] = None, **kwargs: Dict[str, Any]):
...
+            raise DataNodeIsBeingEdited(self.id, self.editor_id)
+
+        if not editor_id and self.edit_in_progress:
+            print("Orchestrator writing without editor_id")
+
         self._write(data)
-        self.track_edit(job_id=job_id, **kwargs)
-        self.unlock_edit()
+        self.last_edit_date = datetime.now()
+        self.edit_in_progress = False  # Ensure it's not locked after writing
```
Comment (jrobinAV) [inline taipy/core/data/data_node.py:462]: The unlock edit that you removed should be reverted. It already sets the `self.edit_in_progress` to False.

File: taipy/core/data/data_node.py [hunk 8]
```diff
@@ -442,9 +444,16 @@ def write(self, data, job_id: Optional[JobId] = None, **kwargs: Dict[str, Any]):
...
 
+        if self.edit_in_progress and editor_id and self.editor_id != editor_id:
+            raise DataNodeIsBeingEdited(self.id, self.editor_id)
+
+        if not editor_id and self.edit_in_progress:
+            print("Orchestrator writing without editor_id")
+
         self._write(data)
-        self.track_edit(job_id=job_id, **kwargs)
-        self.unlock_edit()
```
Comment (jrobinAV) [inline taipy/core/data/data_node.py:453]: PLease revert these two lines.

File: taipy/core/data/data_node.py [hunk 9]
```diff
@@ -426,12 +426,14 @@ def append(self, data, job_id: Optional[JobId] = None, **kwargs: Dict[str, Any])
         """
         from ._data_manager_factory import _DataManagerFactory
 
+        if self.edit_in_progress and self.editor_id != editor_id:
```
Comment (jrobinAV) [inline taipy/core/data/data_node.py:435]: Can you elaborate on why you did not use the same if condition as the `write` function?
What if `editor_id` is `None`?

File: taipy/core/data/data_node.py [hunk 10]
```diff
@@ -426,12 +426,14 @@ def append(self, data, job_id: Optional[JobId] = None, **kwargs: Dict[str, Any])
...
+            raise DataNodeIsBeingEdited(self.id, self.editor_id)
+
         self._append(data)
-        self.track_edit(job_id=job_id, **kwargs)
-        self.unlock_edit()
+        self.track_edit(job_id=job_id, editor_id=editor_id, **kwargs)
```
Comment (jrobinAV) [inline taipy/core/data/data_node.py]:
```suggestion
        self.track_edit(job_id=job_id, editor_id=editor_id, **kwargs)
        self.unlock_edit()
```

File: taipy/core/data/data_node.py [hunk 11]
```diff
@@ -426,12 +426,14 @@ def append(self, data, job_id: Optional[JobId] = None, **kwargs: Dict[str, Any])
         """
         from ._data_manager_factory import _DataManagerFactory
 
+        if self.edit_in_progress and self.editor_id != editor_id:
+            raise DataNodeIsBeingEdited(self.id, self.editor_id)
+
         self._append(data)
-        self.track_edit(job_id=job_id, **kwargs)
-        self.unlock_edit()
```
Comment (jrobinAV) [inline taipy/core/data/data_node.py:437]: Why?

File: taipy/core/data/_file_datanode_mixin.py [hunk 12]
```diff
@@ -141,8 +153,8 @@ def _upload(self, path: str, upload_checker: Optional[Callable[[str, Any], bool]
                 return reason_collection
 
         shutil.copy(upload_path, self.path)
-
-        self.track_edit(timestamp=datetime.now())  # type: ignore[attr-defined]
+        self._write(data)
```
Comment (jrobinAV) [inline taipy/core/data/_file_datanode_mixin.py:156]: The _upload method should not write data directly like that. It is designed to replace the data of a File data node with a new file without passing through the write mechanism.

Please revert this change.

File: taipy/core/data/_file_datanode_mixin.py [hunk 13]
```diff
@@ -100,7 +101,15 @@ def _get_downloadable_path(self) -> str:
...
-    def _upload(self, path: str, upload_checker: Optional[Callable[[str, Any], bool]] = None) -> ReasonCollection:
+    def _upload(
+        self,
+        data: Optional[Any] = None,
+        path: str = "",
+        upload_checker: Optional[Callable[[str, Any], bool]] = None,
+        job_id: Optional["JobId"] = None,
+        editor_id: Optional[str] = None,
+        **kwargs: Dict[str, Any]
+    ) -> None:
```
Comment (jrobinAV) [inline taipy/core/data/_file_datanode_mixin.py:112]: I don't get this change. 

You only need one new argument: an optional `editor_id`. In particular, we don't want data or job_id. We must also return a reason collection!

Why not add a kwargs parameter to pass it to the track edit method, indeed. But this is out of the scope of the issue.

```suggestion
    def _upload(self, path: str, upload_checker: Optional[Callable[[str, Any], bool]] = None, **kwargs: Dict[str, Any]) -> ReasonCollection:
```

File: taipy/core/data/_file_datanode_mixin.py [hunk 14]
```diff
@@ -141,8 +153,8 @@ def _upload(self, path: str, upload_checker: Optional[Callable[[str, Any], bool]
                 return reason_collection
 
         shutil.copy(upload_path, self.path)
-
-        self.track_edit(timestamp=datetime.now())  # type: ignore[attr-defined]
+        self._write(data)
+        self.track_edit(job_id=job_id, editor_id=editor_id, **kwargs)  # type: ignore[attr-defined]
```
Comment (jrobinAV) [inline taipy/core/data/_file_datanode_mixin.py:157]:
```suggestion
        self.track_edit(timestamp=datetime.now(), editor_id=editor_id, **kwargs)  # type: ignore[attr-defined]
```

Comment (jrobinAV) [inline taipy/core/data/_file_datanode_mixin.py:157]: THere is no possible job_id. This cannot be called by the orchestrator.

File: taipy/core/data/_file_datanode_mixin.py [hunk 15]
```diff
@@ -112,6 +121,9 @@ def _upload(self, path: str, upload_checker: Optional[Callable[[str, Any], bool]
         Returns:
             True if the upload was successful, otherwise False.
         """
+        if self.edit_in_progress and self.editor_id != editor_id:
+            raise DataNodeIsBeingEdited(self.id, self.editor_id)
+
```
Comment (jrobinAV) [inline taipy/core/data/_file_datanode_mixin.py:126]: Instead of raising an exception, you should build a specific `Reason` and add it to the `reason_collection`, as it is done below in the code for instance. You can re-use the `DataNodeEditInProgress` reason class.

I would also move this code after the creation of the reason_collection variable: 
` reason_collection = ReasonCollection()`

File: taipy/core/data/in_memory.py [hunk 16]
```diff
@@ -98,3 +97,14 @@ def _read(self):
 
     def _write(self, data):
         in_memory_storage[self.id] = data
+
+    def _append(self, data):
+        """Append data to the existing data in the in-memory storage."""
+        if self.id not in in_memory_storage:
```
Comment (jrobinAV) [inline taipy/core/data/in_memory.py:103]: self.id must not be None. you can remove it from the condition.

File: taipy/core/data/in_memory.py [hunk 17]
```diff
@@ -98,3 +97,14 @@ def _read(self):
...
+            in_memory_storage[self.id] = []
+
+        if not isinstance(in_memory_storage[self.id], list):
+            in_memory_storage[self.id] = [in_memory_storage[self.id]]
```
Comment (jrobinAV) [inline taipy/core/data/in_memory.py:107]: If `in_memory_storage[self.id]` is not a list, a `NotImplementedError` should be raised.

**Review by jrobinAV — COMMENTED** [review]
This is a high-quality PR. Thank you very much.

Please have a look at the various comments.

Can you also create a new test file, `tests/core/data/test_file_datanode_mixin.py,` and test the upload changes you made? The test functions should look like the ones you already created for the write and append functions.

Thx,

**Review by jrobinAV — CHANGES_REQUESTED** [review]
Comments should be resolved first.

**Comment by PrakharJain1509** [comment]
Sure, Give me some time, Will do so in some days.

**Comment by jrobinAV** [comment]
@PrakharJain1509 FYI, this PR is slightly related to your changes. 

https://github.com/Avaiga/taipy/pull/2240

**Comment by jrobinAV** [comment]
@PrakharJain1509 Do you plan to keep working on this PR?
Please let us know if you need any help.

**Comment by jrobinAV** [comment]
PR https://github.com/Avaiga/taipy/pull/2265 already solves the issue.
